producer/cloudwatch_logs.go

```diff
@@ -0,0 +1,164 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package producer
+
+import (
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/cloudwatchlogs"
+	"github.com/trivago/gollum/core"
+	"time"
+)
+
+/*
+AWS CloudWatch specific constants.
+Also see http://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/cloudwatch_limits_cwl.html
+*/
+const (
+	// Maximum number of log events in a batch.
+	maxBatchEvents = 10000
+	// Maximum batch size in bytes.
+	maxBatchSize = 1048576
+	// Maximum event size in bytes.
+	maxEventSize = 262144
+	// A batch of log events in a single PutLogEvents request cannot span more than 24 hours.
+	maxBatchTimeSpan = 24 * time.Hour
+	// How many bytes to append to each log event.
+	eventSizeOverhead = 26
+	// DescribeLogStreams transactions/second.
+	describeLogstreamsDelay = 200 * time.Millisecond
+	// PutLogEvents 5 requests/second/log stream.
+	putLogEventsDelay = 200 * time.Millisecond
+)
+
+type CloudwatchLogs struct {
+	core.BufferedProducer `gollumdoc:"embed_type"`
+	stream                *string
+	group                 *string
+	token                 *string
+	service               *cloudwatchlogs.CloudWatchLogs
+}
+
+func init() {
+	core.TypeRegistry.Register(CloudwatchLogs{})
+}
+
+// Configure initializes this producer with values from a plugin config.
+func (prod *CloudwatchLogs) Configure(conf core.PluginConfigReader) error {
+	prod.SetStopCallback(prod.flush)
+	stream := conf.GetString("stream", "")
```

Comment:
Consider using struct tags for stream and group (non-pointer) and only leave the checks on empty in the config method